### ISSUE TITLE
[Perf] Enable torch.compile for S2-Pro Slow AR backbone + Fast AR codebook decoder

### ISSUE THREAD

**Issue by Ccyest** [sgl-project/sglang-omni#571]
## Motivation

S2-Pro currently runs with CUDA graph capture enabled on the Slow AR path but with zero `torch.compile` anywhere. The Qwen3 backbone (24 layers) runs fully eager kernels inside the captured graph, and the Fast AR codebook decoder (`_decode_codebooks`: constrained semantic sampling + 9-step `forward_kvcached` loop) is also entirely eager.

Benchmarking from PR #153 (@Ratish1) already quantified the opportunity on this model:

| Configuration | Startup time | Steady-state throughput | Δ throughput |
|---|---|---|---|
| CUDA graph only (current) | ~33s | 88 tok/s | baseline |
| Partial compile (fast head only) | ~54s | 121 tok/s | **+37%** |
| Full-model compile | ~137s | 126 tok/s | +43% |

Partial compile captures ~85% of the total compile gain at ~20% of the startup cost. The Fast AR codebook decoder is the primary compile target — its tight loop of small matmuls, embedding lookups, and argmax steps benefits most from kernel fusion.

## Architecture context

S2-Pro uses a Dual-AR architecture:

```
Slow AR (Qwen3 24L backbone, RadixAttention, CUDA-graph-captured)
  ↓ hidden_states
_decode_codebooks:
  ├── Constrained semantic sampling (top-k/top-p with RAS + repetition penalty)
  ├── project_in → forward_kvcached(codebook_idx=0)
  └── for cb_idx in 1..9:
        forward_kvcached → argmax → embeddings → next step
  ↓ output_codes [bs, num_codebooks+1]
```

The Fast AR loop runs 9 sequential `audio_decoder.forward_kvcached()` calls per Slow AR step. Each call is a small 4-layer transformer forward — individually too small to saturate the GPU, but fuseable by torch.compile into a much more efficient kernel sequence.

## Scope

- Model: `sglang_omni/models/fishaudio_s2_pro/sglang_model.py` (`S2ProSGLangTextModel`)
- Model runner: `sglang_omni/models/fishaudio_s2_pro/model_runner.py` (`FishS2ProModelRunner`)
- Stage factory: `sglang_omni/models/fishaudio_s2_pro/stages.py` (`create_sglang_tts_engine_executor`)
- Fast AR head: `audio_decoder` (loaded in `bootstrap.py`, attached via `setup_vq_decode`)

Two compile scopes to pursue, in order:

### Scope 1: Partial compile (Fast AR codebook decoder)

1. Extract `_decode_codebooks` into a `fullgraph=True`-compatible standalone function with pure tensor interface (no `self._output_codes[:bs]` slicing inside the compiled region)
2. Apply `torch.compile(mode="max-autotune-no-cudagraphs", fullgraph=True)` to the extracted function
3. Expected result: ~121 tok/s (+37%), ~54s startup

### Scope 2: Full-model compile (Slow AR + Fast AR)

1. Pass `enable_torch_compile: true` through `server_args_overrides` so sglang applies torch.compile to the full `forward()`
2. Expected result: ~126 tok/s (+43%), ~137s startup (mitigable by inductor cache)

Scope 1 is the priority — it captures the majority of the gain with minimal startup cost.

## Baseline

| Metric | Config | Value |
|---|---|---|
| Throughput (tok/s) | CUDA graph only, H200, bs=1 | 88 tok/s (PR #153) |
| RTF | single H200, C=1 | 0.34 (README) |
| Startup | CUDA graph only | ~33s |

Re-measure on the current main before starting — the numbers from PR #153 may have shifted with subsequent changes.

## Acceptance Criteria

**Accuracy (no regression)**
- [ ] SeedTTS EN WER: no regression vs eager baseline
- [ ] SeedTTS EN speaker similarity: no regression vs eager baseline
- [ ] Deterministic codebook outputs match eager path for fixed seed

**Performance (must improve)**
- [ ] Partial compile: throughput ≥ 115 tok/s on H200 (≥ +30% over CUDA-graph-only baseline)
- [ ] Startup time: partial compile ≤ 60s
- [ ] Full compile (if pursued): throughput ≥ 120 tok/s

**Code health**
- [ ] No `torch.compile()` calls inside model files — compile applied at framework/factory level
- [ ] Existing unit tests still pass
- [ ] `fullgraph=True` — zero graph breaks in the compiled path

## Known Pitfalls

**`torch.multinomial` in `_decode_codebooks`.** Semantic token sampling uses `torch.multinomial(probs, num_samples=1)`. This may cause a graph break under `fullgraph=True`. If it does, either:
- Split the compile boundary: compile the logit processing (bias, penalty, top-k/top-p masking, softmax) separately from the sampling step
- Or replace with a manual Gumbel-max trick (`argmax(log(probs) - log(-log(uniform)))`) which is compile-friendly

**Sequential codebook loop.** The 9-step `for cb_idx in range(1, num_codebooks)` loop with `forward_kvcached` is inherently sequential (each step depends on the previous argmax). torch.compile can fuse within each iteration but not across iterations. Unrolling the loop may help the compiler but increases code size. Start with the natural loop; unroll only if profiling shows the loop overhead matters.

**`audio_decoder.reset_caches()` side effect.** Called at the start of `_decode_codebooks`. If this mutates state that torch.compile captures as a constant at trace time, the reset will be silently skipped on subsequent calls. Verify that cache reset works correctly under compile by checking that codebook outputs change across decode steps (they should, since each step has different hidden_states input).

**VQ embedding injection in Slow AR forward.** During decode, `forward()` combines text embeddings with VQ codebook embeddings via `torch.where(vq_mask, combined, hidden_states)`. This is graph-safe (static tensor shapes, elementwise ops), but if `enable_torch_compile` is applied to the full model, verify that the `_vq_codebook_embeddings` lookup with offset arithmetic compiles correctly.

**FlashAttention 3 backend.** S2-Pro forces `attention_backend = "fa3"` to match training-time numerics. Verify that FA3 is compatible with torch.compile's tracing — if the backend uses custom CUDA kernels not registered with torch.compile, they may appear as opaque calls and prevent `fullgraph=True` on Scope 2.

**Deferred CUDA graph capture.** `stages.py` temporarily sets `disable_cuda_graph = True` during `create_sglang_infrastructure`, then re-enables and calls `init_device_graphs()` after `setup_vq_decode` attaches the audio decoder. torch.compile must be applied before graph capture so the compiled kernels are what gets captured. Verify the ordering: compile → attach decoder → capture graphs.

## Suggested Approach

1. **Re-measure baseline.** Run SeedTTS eval with the current main on H200. Record tok/s, RTF, WER, startup time.
2. **Scope 1: Partial compile of `_decode_codebooks`.**
   - Extract the core logic into a pure-tensor function
   - Apply `torch.compile(fullgraph=True, mode="max-autotune-no-cudagraphs")`
   - Check for graph breaks with `TORCH_LOGS=graph_breaks`
   - If `torch.multinomial` breaks the graph, split the compile boundary around it
   - Benchmark: expect ~121 tok/s
3. **Verify accuracy.** Run SeedTTS EN eval, compare WER and speaker similarity.
4. **Scope 2 (optional): Full-model compile.**
   - Add `"enable_torch_compile": True, "torch_compile_max_bs": 32` to `server_args_overrides`
   - Test if sglang's compile + CUDA graph coexist with S2-Pro's deferred graph capture pattern
   - Benchmark: expect marginal gain over Scope 1 (~126 vs ~121 tok/s)
5. **Inductor cache.** Set `TORCHINDUCTOR_CACHE_DIR` to persist compiled artifacts across restarts, reducing subsequent startup to near-baseline.

## References

- PR #153 — torch.compile benchmark data for S2-Pro (@Ratish1)
- Issue #172 — framework-level torch.compile design (covers all models)
- S2-Pro model: `sglang_omni/models/fishaudio_s2_pro/sglang_model.py`
- Fast AR head: loaded in `sglang_omni/models/fishaudio_s2_pro/bootstrap.py` (`load_audio_decoder`)
- S2-Pro architecture: `sglang_omni/models/fishaudio_s2_pro/README.md`
- Benchmark entry point: `benchmarks/eval/benchmark_tts_seedtts.py`

**Comment by nagisa-kunhah** [comment]
@Ccyest Hi, I noticed that #239 already implements the Phase 1 framework-level torch.compile support for the S2-Pro codebook decoder. Should we wait for #239 to be merged before working on this issue, or is there any remaining scope from #571 that can be worked on independently?

**Comment by zhaochenyang20** [comment]
@nagisa-kunhah you can fork from #239 branch, and continue working on it. Turn all the optimzation for S2 Pro default on, and pass the CI. Thanks!
